File: packages/typescript/tests/parts.test.ts
import { describe, it, expect } from 'vitest';
import { FilePart, DataPart, Artifact, textPart, parsePart } from '../src/parts.js';

describe('textPart', () => {
  it('should create a text part', () => {
    const part = textPart('Hello, world!');
    expect(part.type).toBe('text');
    expect(part.text).toBe('Hello, world!');
  });
});

describe('FilePart', () => {
  it('should create with bytes', () => {
    const part = new FilePart({
      name: 'test.txt',
      mimeType: 'text/plain',
      data: Buffer.from('Hello, world!'),
    });

    expect(part.name).toBe('test.txt');
    expect(part.isBytes).toBe(true);
    expect(part.isUri).toBe(false);
  });

  it('should create with string data', () => {
    const part = new FilePart({
      name: 'test.txt',
      data: 'Hello, world!',
    });

    expect(part.isBytes).toBe(true);
    expect(part.data?.toString()).toBe('Hello, world!');
  });

  it('should create with URI', () => {
    const part = new FilePart({
      name: 'remote.txt',
      uri: 'https://example.com/file.txt',
    });

    expect(part.isUri).toBe(true);
    expect(part.isBytes).toBe(false);
  });

  it('should convert to A2A v1.0 format with bytes (raw)', () => {
    const part = new FilePart({
      name: 'test.txt',
      mimeType: 'text/plain',
      data: Buffer.from('Hello'),
    });

    const a2a = part.toA2A();

    expect(a2a.raw).toBe(Buffer.from('Hello').toString('base64'));
    expect(a2a.filename).toBe('test.txt');
    expect(a2a.mediaType).toBe('text/plain');
  });

  it('should convert to A2A v1.0 format with URI (url)', () => {
    const part = new FilePart({
      name: 'test.txt',
      uri: 'https://example.com/file.txt',
    });

    const a2a = part.toA2A();

    expect(a2a.url).toBe('https://example.com/file.txt');
    expect(a2a.filename).toBe('test.txt');
  });

  it('should parse from A2A v1.0 format', () => {
    const a2aData = {
      raw: Buffer.from('Hello').toString('base64'),
      filename: 'test.txt',
      mediaType: 'text/plain',
    };

    const part = FilePart.fromA2A(a2aData);

    expect(part.name).toBe('test.txt');
    expect(part.mimeType).toBe('text/plain');
    expect(part.data?.toString()).toBe('Hello');
  });

  it('should read bytes', async () => {
    const part = new FilePart({
      name: 'test.txt',
      data: 'Hello',
    });

    const content = await part.readBytes();
    expect(content.toString()).toBe('Hello');
  });

  it('should read text', async () => {
    const part = new FilePart({
      name: 'test.txt',
      data: 'Hello, world!',
    });

    const content = await part.readText();
    expect(content).toBe('Hello, world!');
  });
});

describe('DataPart', () => {
  it('should create with data', () => {
    const part = new DataPart({ key: 'value', count: 42 });
    expect(part.data).toEqual({ key: 'value', count: 42 });
  });

  it('should convert to A2A v1.0 format', () => {
    const part = new DataPart({ key: 'value' });
    const a2a = part.toA2A();

    expect(a2a.data).toEqual({ key: 'value' });
    expect(a2a.mediaType).toBe('application/json');
  });

  it('should parse from A2A v1.0 format', () => {
    const part = DataPart.fromA2A({ data: { key: 'value' } });
    expect(part.data).toEqual({ key: 'value' });
  });
});

describe('Artifact', () => {
  it('should create empty artifact', () => {
    const artifact = new Artifact({ name: 'report.json' });
    expect(artifact.name).toBe('report.json');
    expect(artifact.parts).toEqual([]);
  });

  it('should add text', () => {
    const artifact = new Artifact();
    artifact.addText('Hello');

    expect(artifact.parts).toHaveLength(1);
    expect((artifact.parts[0] as any).text).toBe('Hello');
  });

  it('should add file', () => {
    const artifact = new Artifact();
    const file = new FilePart({ name: 'test.txt', data: 'content' });
    artifact.addFile(file);

    expect(artifact.parts).toHaveLength(1);
    expect((artifact.parts[0] as any).name).toBe('test.txt');
  });

  it('should add data', () => {
    const artifact = new Artifact();
    artifact.addData({ key: 'value' });

    expect(artifact.parts).toHaveLength(1);
    expect((artifact.parts[0] as any).data).toEqual({ key: 'value' });
  });

  it('should support chaining', () => {
    const artifact = new Artifact({ name: 'combined' }).addText('Summary').addData({ count: 10 });

    expect(artifact.parts).toHaveLength(2);
  });

  it('should convert to A2A format', () => {
    const artifact = new Artifact({ name: 'report', description: 'Test' });
    artifact.addText('Hello');

    const a2a = artifact.toA2A();

    expect(a2a.name).toBe('report');
    expect(a2a.description).toBe('Test');
    expect(a2a.parts as any[]).toHaveLength(1);
  });
});

describe('parsePart', () => {
  it('should parse text part', () => {
    const part = parsePart({ text: 'Hello' });
    expect(part.type).toBe('text');
    expect((part as any).text).toBe('Hello');
  });

  it('should parse raw file part', () => {
    const part = parsePart({
      raw: Buffer.from('data').toString('base64'),
      filename: 'test.txt',
      mediaType: 'text/plain',
    });
    expect(part.type).toBe('file');
  });

  it('should parse url file part', () => {
    const part = parsePart({ url: 'https://example.com/f.txt', filename: 'f.txt' });
    expect(part.type).toBe('file');
    expect((part as any).isUri).toBe(true);
  });

  it('should parse data part', () => {
    const part = parsePart({ data: { key: 'value' } });
    expect(part.type).toBe('data');
  });
});

File: packages/java/src/test/java/com/a2alite/PartsTest.java
package com.a2alite;

import com.a2alite.parts.Artifact;
import com.a2alite.parts.DataPart;
import com.a2alite.parts.FilePart;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Base64;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class PartsTest {

    // === FilePart Tests ===

    @Test
    void shouldCreateFilePartFromBytes() {
        var data = "hello world".getBytes();
        var file = new FilePart("test.txt", "text/plain", data);

        assertThat(file.getName()).isEqualTo("test.txt");
        assertThat(file.getMimeType()).isEqualTo("text/plain");
        assertThat(file.isBytes()).isTrue();
        assertThat(file.isUri()).isFalse();
    }

    @Test
    void shouldCreateFilePartFromPath(@TempDir Path tempDir) throws IOException {
        var path = tempDir.resolve("test.txt");
        Files.writeString(path, "file content");

        var file = FilePart.fromPath(path);
        assertThat(file.getName()).isEqualTo("test.txt");
        assertThat(file.getMimeType()).isEqualTo("text/plain");
        assertThat(file.isBytes()).isTrue();
    }

    @Test
    void shouldCreateFilePartFromPathWithMimeType(@TempDir Path tempDir) throws IOException {
        var path = tempDir.resolve("data.bin");
        Files.write(path, new byte[]{1, 2, 3});

        var file = FilePart.fromPath(path, "application/custom");
        assertThat(file.getMimeType()).isEqualTo("application/custom");
    }

    @Test
    void shouldCreateFilePartFromUri() {
        var file = FilePart.fromUri("https://example.com/file.pdf", "file.pdf");
        assertThat(file.isUri()).isTrue();
        assertThat(file.isBytes()).isFalse();
        assertThat(file.getUri()).isEqualTo("https://example.com/file.pdf");
    }

    @Test
    void shouldReadBytesFromData() throws Exception {
        var data = "hello".getBytes();
        var file = new FilePart("test.txt", "text/plain", data);
        assertThat(file.readBytes()).isEqualTo(data);
    }

    @Test
    void shouldReadTextFromData() throws Exception {
        var file = new FilePart("test.txt", "text/plain", "hello world".getBytes());
        assertThat(file.readText()).isEqualTo("hello world");
    }

    @Test
    void shouldThrowWhenNoDataOrUri() {
        var file = new FilePart("empty.txt", "text/plain", null, null);
        assertThatThrownBy(file::readBytes)
            .isInstanceOf(IllegalStateException.class)
            .hasMessageContaining("no data or URI");
    }

    @Test
    void shouldConvertBytesToA2A() {
        var data = "hello".getBytes();
        var file = new FilePart("test.txt", "text/plain", data);
        var a2a = file.toA2A();

        assertThat(a2a.get("type")).isEqualTo("file");
        assertThat(a2a.get("kind")).isEqualTo("file");

        @SuppressWarnings("unchecked")
        var fileMap = (Map<String, Object>) a2a.get("file");
        assertThat(fileMap.get("name")).isEqualTo("test.txt");
        assertThat(fileMap.get("mimeType")).isEqualTo("text/plain");
        assertThat(fileMap.get("bytes")).isEqualTo(Base64.getEncoder().encodeToString(data));
    }

    @Test
    void shouldConvertUriToA2A() {
        var file = FilePart.fromUri("https://example.com/f.pdf", "f.pdf");
        var a2a = file.toA2A();

        @SuppressWarnings("unchecked")
        var fileMap = (Map<String, Object>) a2a.get("file");
        assertThat(fileMap.get("uri")).isEqualTo("https://example.com/f.pdf");
        assertThat(fileMap).doesNotContainKey("bytes");
    }

    @Test
    void shouldParseFromA2AWithBytes() throws Exception {
        var encoded = Base64.getEncoder().encodeToString("hello".getBytes());
        var a2a = Map.<String, Object>of(
            "file", Map.of(
                "name", "test.txt",
                "mimeType", "text/plain",
                "bytes", encoded
            )
        );

        var file = FilePart.fromA2A(a2a);
        assertThat(file.getName()).isEqualTo("test.txt");
        assertThat(file.getMimeType()).isEqualTo("text/plain");
        assertThat(file.isBytes()).isTrue();
        assertThat(new String(file.readBytes())).isEqualTo("hello");
    }

    @Test
    void shouldParseFromA2AWithUri() {
        var a2a = Map.<String, Object>of(
            "file", Map.of(
                "name", "doc.pdf",
                "uri", "https://example.com/doc.pdf"
            )
        );

        var file = FilePart.fromA2A(a2a);
        assertThat(file.isUri()).isTrue();
        assertThat(file.getUri()).isEqualTo("https://example.com/doc.pdf");
    }

    @Test
    void shouldGuessMimeTypes() throws IOException {
        assertThat(FilePart.fromPath(createTempFile("test.json", "{}")).getMimeType())
            .isEqualTo("application/json");
        assertThat(FilePart.fromPath(createTempFile("img.png", "x")).getMimeType())
            .isEqualTo("image/png");
        assertThat(FilePart.fromPath(createTempFile("doc.pdf", "x")).getMimeType())
            .isEqualTo("application/pdf");
        assertThat(FilePart.fromPath(createTempFile("style.css", "x")).getMimeType())
            .isEqualTo("text/css");
    }

    @Test
    void shouldDefaultMimeType() {
        var file = new FilePart("unknown.xyz", null, new byte[0]);
        assertThat(file.getMimeType()).isEqualTo("application/octet-stream");
    }

    // === DataPart Tests ===

    @Test
    void shouldCreateDataPart() {
        var data = new DataPart(Map.of("count", 42, "name", "test"));
        assertThat(data.getData()).containsEntry("count", 42);
        assertThat(data.getData()).containsEntry("name", "test");
    }

    @Test
    void shouldGetByKey() {
        var data = new DataPart(Map.of("count", 42));
        Integer count = data.get("count");
        assertThat(count).isEqualTo(42);
    }

    @Test
    void shouldConvertDataPartToA2A() {
        var data = new DataPart(Map.of("key", "value"));
        var a2a = data.toA2A();
        assertThat(a2a.get("type")).isEqualTo("data");
        assertThat(a2a.get("kind")).isEqualTo("data");
        assertThat(a2a.get("data")).isEqualTo(Map.of("key", "value"));
    }

    @Test
    void shouldParseDataPartFromA2A() {
        var a2a = Map.<String, Object>of("data", Map.of("key", "value"));
        var data = DataPart.fromA2A(a2a);
        assertThat(data.getData()).containsEntry("key", "value");
    }

    @Test
    void shouldHandleEmptyDataPart() {
        var a2a = Map.<String, Object>of();
        var data = DataPart.fromA2A(a2a);
        assertThat(data.getData()).isEmpty();
    }

    // === Artifact Tests ===

    @Test
    void shouldBuildArtifact() {
        var artifact = new Artifact("report", "A report")
            .addText("Summary here")
            .addData(Map.of("count", 42));

        assertThat(artifact.getName()).isEqualTo("report");
        assertThat(artifact.getDescription()).isEqualTo("A report");
        assertThat(artifact.getParts()).hasSize(2);
    }

    @Test
    void shouldAddFilePart() {
        var file = new FilePart("data.csv", "text/csv", "a,b,c".getBytes());
        var artifact = new Artifact("export").addFile(file);
        assertThat(artifact.getParts()).hasSize(1);
    }

    @Test
    void shouldConvertArtifactToA2A() {
        var artifact = new Artifact("test")
            .addText("hello")
            .addData(Map.of("x", 1));

        var a2a = artifact.toA2A();
        assertThat(a2a.get("name")).isEqualTo("test");
        assertThat(a2a.get("parts")).isNotNull();
    }

    @Test
    void shouldSetMetadata() {
        var artifact = new Artifact("test")
            .setMetadata(Map.of("version", "1.0"));
        assertThat(artifact.getMetadata()).containsEntry("version", "1.0");
    }

    // Helper
    private Path createTempFile(String name, String content) throws IOException {
        var dir = Files.createTempDirectory("a2a-test");
        var file = dir.resolve(name);
        Files.writeString(file, content);
        return file;
    }
}
